book/invalidation.md

```diff
@@ -1926,24 +1926,39 @@ class ProtectedField:
 ```
 
 With these changes, the descendant dirty bits should now be set
-correctly, and the `layout_needed` approach above should work. You can
-replicate this on all the other layout types. On `DocumentLayout`, you
-need to be a little careful about the order because it receives the
-frame width and zoom level as arguments:
+correctly, and the `layout_needed` approach above should work as long
+as you include all of the protected fields for each layout type. In
+`DocumentLayout`, you do need to be a little careful, since it
+receives the frame width and zoom level as an argument. You need to
+make sure to `mark` those fields if they changed. The `width` changes
+when the `frame_width` changes, here:
 
 ``` {.python}
-class DocumentLayout:
-    def layout(self, width, zoom):
-        self.zoom.set(zoom)
-        self.width.set(width - 2 * device_px(HSTEP, zoom))
-        if not self.layout_needed(): return
+class IframeLayout(EmbedLayout):
+    def layout(self):
+        if self.node.frame:
+            # ...
+            self.node.frame.document.height.mark()
+            self.node.frame.set_needs_render()
```

Comment:
why is set_needs_render needed here?

Reply:
So, it at least needs to `set_needs_layout`, right?